src/lib/bootloader/grub_install.rb

```diff
@@ -2,30 +2,73 @@
 require "yast2/execute"
 
 Yast.import "Arch"
+Yast.import "Report"
 
 module Bootloader
   # Wraps grub install script for easier usage.
   class GrubInstall
+    include Yast::Logger
+    include Yast::I18n
+
     def initialize(efi: false)
       @efi = efi
+      textdomain "bootloader"
     end
 
+    # Runs grub2 install command.
+    #
+    # @param devices[Array<String>] list of devices where grub2 should be installed.
+    #   Ignored when grub2 does not need device.
+    # @param secure_boot [Boolean] if secure boot variant should be used
+    # @param trusted_boot [Boolean] if trusted boot variant should be used
+    # @return [Array<String>] list of devices for which install failed
     def execute(devices: [], secure_boot: false, trusted_boot: false)
       raise "cannot have secure boot without efi" if secure_boot && !efi
 
       cmd = basic_cmd(secure_boot, trusted_boot)
 
       if no_device_install?
         Yast::Execute.on_target(cmd)
+        []
       else
-        devices.each { |d| Yast::Execute.on_target(cmd + [d]) }
+        return [] if devices.empty?
+
+        last_failure = nil
+        res = devices.select do |device|
+          begin
+            Yast::Execute.on_target!(cmd + [device])
```

Comment:
just for record: what about shell escaping ```device``` ?

Reply:
Cheetah is not affected by shell execution as it does not use shell and pass arguments directly. See https://github.com/openSUSE/cheetah#features

Reply:
ok, thanks. I will have to study cheetah more deeply.